eth_abi/registry.py

```diff
@@ -0,0 +1,131 @@
+from cytoolz import curry
+
+from .utils.parsing import collapse_type
+
+
+class ABIRegistryError(Exception):
+    pass
+
+
+class NoEntryFound(ABIRegistryError):
+    pass
+
+
+class PredicateMapping:
+    """
+    Acts as a mapping from predicate functions to values.  Values are retrieved when their
+    corresponding predicate matches a set of given inputs.  Predicates can also be labeled to
+    facilitate removal from the mapping.
+    """
+    def __init__(self):
+        self._values = {}
+        self._labeled_predicates = {}
+
+    def add(self, predicate, value, label=None):
+        if predicate in self._values:
+            raise ValueError('Predicate already exists')
+
+        if label is not None:
+            if label in self._labeled_predicates:
+                raise ValueError('Predicate with label "{}" already exists'.format(label))
+
+            self._labeled_predicates[label] = predicate
+
+        self._values[predicate] = value
+
+    def find(self, *args, **kwargs):
+        results = tuple(
+            value for predicate, value in self._values.items()
+            if predicate(*args, **kwargs)
```

Comment:
*nitpick*

I think that this tuple comprehension is equivalent to the [`toolz.keyfilter`](http://toolz.readthedocs.io/en/latest/api.html#toolz.dicttoolz.keyfilter).  Always questionable whether using such functions is an improvement as they are often a readability trade off.  Merely pointing it out so you know the tool exists.